mir/src/passes/unrolling/match_optimizer.rs

```diff
@@ -0,0 +1,241 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    ops::Deref,
+};
+
+use miden_diagnostics::SourceSpan;
+
+use crate::{
+    CompileError,
+    ir::{
+        Add, ConstantValue, Link, MatchArm, MirValue, Mul, Op, Parent, QuadFelt, RandomInputs,
+        SpannedMirValue, Sub, Value,
+    },
+    passes::duplicate_node,
+};
+
+type ConstraintEvaluationMap = BTreeMap<usize, Vec<(Link<Op>, Link<Op>)>>;
+
+/// This struct provides methods used to combine and optimize constraints contained in match
+/// statements (corresponding to `If` nodes in the MIR)
+#[derive(Debug)]
+pub struct MatchOptimizer<'a> {
+    // A reference to the random inputs used to evaluate the constraints
+    // (common for all matches)
+    random_inputs: &'a mut RandomInputs,
+    // Used to keep track of the evaluations values we've seen so far
+    // in the current `If` node.
+    node_evals: Vec<QuadFelt>,
+    // Used to keep track of the evaluation of each constraint
+    // We use indices to the node_evals vector as keys, to avoid non-determinism for
+    // iterating across random evaluations.
+    constraints_evaluation_indices: ConstraintEvaluationMap,
+}
+
+impl<'a> MatchOptimizer<'a> {
+    /// Instantiates a new `MatchOptimizer` with the given `RandomInputs`.
+    pub fn new(random_inputs: &'a mut RandomInputs) -> Self {
+        Self {
+            random_inputs,
+            node_evals: Vec::new(),
+            constraints_evaluation_indices: ConstraintEvaluationMap::new(),
+        }
+    }
+
+    /// For a given match arm, splits main and bus-related constraints, and evaluates the main
+    /// constraints
+    pub fn evaluate_match_arm(
+        &mut self,
+        match_arm: &MatchArm,
+    ) -> Result<Vec<Link<Op>>, CompileError> {
```

Comment:
I would also mention that:

- We return a list of bus-related constraints (and it seems like we may be transforming them as well?).
- That we group all main constraints by the evaluations - i.e., two constraints that evaluate to the same value, will be put in the same group.